napari/util/keybindings.py

```diff
@@ -0,0 +1,582 @@
+
+
+class BindOperation(Enum):
+    BIND = 'bind'
+    UNBIND = 'unbind'
+    REBIND = 'rebind'
+
+
+def _expand_shorthand(shorthand_seq):
+    """Expand shorthand form of a key sequence.
+
+    C -> Control
+    M -> Alt
+    S -> Shift
+    s -> Meta
+    """
+    return re.sub(shorthand_modifier_patt,
+                  lambda m: MODIFIER_KEYS[m.group(0)].name,
+                  shorthand_seq)
+
+
+def parse_seq(seq):
+    """Parse a key sequence into its components in a comparable format.
+
+    Parameters
+    ----------
+    seq : str
+        Key sequence.
+
+    Returns
+    -------
+    key : str
+        Base key of the sequence.
+    modifiers : set of str
+        Modifier keys of the sequence.
+    """
+    parsed = re.split('-(?=.+)', seq)
+    modifiers, key = parsed[:-1], parsed[-1]
+
+    return key, set(modifiers)
+
+
+def components_to_seq(key, modifiers):
+    """Combine components to become a key sequence.
+
+    Modifier keys will always be combined in the same order:
+    Control, Alt, Shift, Meta
+
+    Shift pressed with any letter will be consumed
+    to transform it to upper case, e.g. Shift + w => W.
+    Otherwise, the letter will be lower case.
+    Shift will also be consumed without transformation when pressed
+    with any other non-special key unless Control is also a modifier.
+
+    Parameters
+    ----------
+    key : str or vispy.app.Key
+        Base key.
+    modifiers : sequence of str or vispy.app.Key
+        Modifier keys.
+
+    Returns
+    -------
+    seq : str
+        Generated key sequence.
+    """
+    if len(key) == 1 and key.isalpha():  # it's a letter
+        if 'Shift' not in modifiers:
+            key = key.lower()
+            cond = lambda m: True  # noqa: E731
+        else:
+            # Shift is consumed to create upper-case letter
+            key = key.upper()
+            cond = lambda m: m != 'Shift'  # noqa: E731
+    elif key in SPECIAL_KEYS:
+        # remove redundant information i.e. an output of 'Shift-Shift'
+        cond = lambda m: m != key  # noqa: E731
+    else:
+        # Shift is consumed to transform key
+
+        # bug found on OSX: Command will cause Shift to not
+        # transform the key so do not consume it
+        # note: 'Control' is OSX Command key
+        cond = lambda m: m != 'Shift' or 'Control' in modifiers  # noqa: E731
+
+    modifiers = tuple(key.name for key in
+                      filter(lambda key: key in modifiers and cond(key),
+                             MODIFIER_KEYS.values()))
+
+    return '-'.join(modifiers + (key,))
+
+
+def normalize_key_sequence(seq):
+    """Normalize key sequence to make it easily comparable.
+
+    All aliases are converted and modifier orders are fixed to:
+    Control, Alt, Shift, Meta
+
+    Parameters
+    ----------
+    seq : str
+        Key sequence.
+
+    Shift pressed with any letter will be consumed
+    to transform it to upper case, e.g. Shift + w => W.
```

Comment:
Can you write a bit more about the shift-consuming choice? Is that standard? I kind of dislike it because some people might read `W` as just the key `w`...

Reply:
this is tricky, most of the menu bars that I see just use capital letters and then the up arrow if the shift key is to be pressed, but i think using `w` and `W` also makes sense

Reply:
> most of the menu bars that I see just use capital letters and then the up arrow if the shift key is to be pressed

Exactly. I think expressing the shortcut with the capital is kinda non-standard and therefore bad...

Reply:
@freeman-lab want to weigh in here?

Reply:
this was more to play it in-line with how Qt handles other key combinations

for example, pressing `Shift+2` actually gets read in by Qt as `Shift+@`

Reply:
I think you mean `Shift-2` gets read-in as `@`?

At any rate, we should aim for our user-facing documentation to be as simple and intuitive as possible, and not have it bound to implementation details. In this case, I think writing "Shift-2" as the shortcut is way more intuitive than writing `@` as the shortcut.

Reply:
no, `Shift-2` gets read in as `@`

there is absolutely no way for us to know that `2` was ever pressed due to regional keyboard differences

which is why `Shift-2` as the shortcut would be invalid; also if we did want that we would have to create a table of all of the different key combos and their unicode and convert them back (which again would only work on US keyboards)

Reply:
Fascinating! Yes, `@` is AltGr + 2 on ISO keyboards. What a mess.

Reply:
@kne42 I think the "right" way to do this is to set/display the key combo we want, e.g. "shift-2", and then process the key based on the OS/system keyboard setting. This might be super hard but there might be pre-existing libraries to deal with all this stuff?

Reply:
seems way too complex to me. i think it's best that we leave this alone for now